Add repository credential management API and CLI (addresses #2136)

jannfis opened the pull request:
This PR adds repository credential management to ArgoCD, as discussed in #2136. It changes the following:

- ~~Add the following CRUD operations for repository credentials in `RepositoryService` API~~ Introduces new `RepoCredsService` API which implements the following CRUD operations for managing repository credentials:
  - `ListRepositoryCredentials()`
  - `CreateRepositoryCredentials()`
  - `UpdateRepositoryCredentials()`
  - `DeleteRepositoryCredentials()`

- Deprecation of the following CRUD operations for repository management in `RepositoryService` API. The deprecated methods still available for backwards compatibility, but I think should be safe to remove with the release of v2.0:
  - `List()` deprecated in favour of `ListRepositories()`
  - `Create()` deprecated in favour of `CreateRepository()`
  - `Update()` deprecated in favour of `UpdateRepository()`
  - `Delete()` deprecated in favour of `DeleteRepository()`

- ~~Changes the semantics of `argocd repo` sub-commands by introducing the `--creds` switch to sub-commands `list`, `add` and `rm` to perform operations on repository credentials instead of repositories.~~

- Introduces the `repocreds` sub-command to CLI, which can be used to manage repository credential templates. It mimics `repo` sub-command to some extent and supports the `add`, `rm` and `list` operations, with `--upsert` support.

- Introduces ~~`--force-refresh`~~ `--refresh <mode>` switch to `argocd repo list` command and accompanying`refresh` field in `RepoQuery` message of `RepositoryService` API  to force a cache refresh of the listed repository's connection status

- _Update 08/26/19_: Changes behaviour of credential lookup from "first match" to "best match", i.e. given the following two cred URLs `https://reposerver/repos`and `https://reposerver/repos/private`, credential lookup for repository `https://reposerver/repos/private/myprivate` would return credentials configured for `https://reposerver/repos/private`, not `https://reposerver/repos`

- _Update 08/26/19 2_: Adds initial functionality to the web UI for listing, adding and removing repository credential templates - all built into the 'Repositories' settings page. Also added ability to force a cache reload of repositories with a new button on top of the page. See screenshots.

- Update 08/27/19: Added documentation for Operator Manual (declarative) and User Manual (Web UI and CLI)

- Update 09/28/19: Introduce new `repocreds` sub-command to CLI and removes `--creds` switch (and accompanying functionality) from `repo` sub-command (docs still need updating).

![image](https://user-images.githubusercontent.com/3942683/63721827-5aef0c80-c852-11e9-9288-96ba66e9692c.png)

![image](https://user-images.githubusercontent.com/3942683/63721897-7c4ff880-c852-11e9-9f4f-d47081993b46.png)

![image](https://user-images.githubusercontent.com/3942683/63721966-9689d680-c852-11e9-97d7-89a96e548fea.png)


CLI examples:

```
[eris@docker-build argo-cd]$ ./dist/argocd repocreds --list
URL_PATTERN                 USERNAME  SSH_CREDS  TLS_CREDS
https://docker-build/repos  test      false      false
[eris@docker-build argo-cd]$ ./dist/argocd repo list
REPO                                            INSECURE  LFS    USER  STATUS      MESSAGE
https://docker-build/repos/argocd-example-apps  false     false  test  Successful
[eris@docker-build argo-cd]$ ./dist/argocd repocreds rm https://docker-build/repos
[eris@docker-build argo-cd]$ ./dist/argocd repo list --refresh hard
REPO                                            INSECURE  LFS    USER  STATUS  MESSAGE
https://docker-build/repos/argocd-example-apps  false     false  -     Failed  Unable to connect to repository: authentication required
```

alexec commented:
looking good so far - I've some polish comments - but I'll wait until this is `Ready For Review`

jannfis commented:
I think this PR is complete now. Sorry for it being so large again, I owe you guys a beer or two for all the review work I generate.

jannfis commented on cmd/argocd/commands/repo.go:
```diff
@@ -72,6 +75,18 @@ Add a HTTPS repository using username/password without verifying the server's TL
+			// If user specified --creds, error out on invalid parameters
+			if credsOnly {
+				if insecureIgnoreHostKey || insecureSkipServerVerification {
+					err := fmt.Errorf("--insecure-skip-server-verification and --insecure-ignore-host-key cannot be specified with --creds")
```
See additional comment about `insecureIgnoreHostKey` below.

jannfis commented on cmd/argocd/commands/repo.go:
```diff
@@ -195,7 +238,20 @@ func printRepoTable(repos []appsv1.Repository) {
+// Print the repository credentials as table
+func printRepoCredsTable(repos []appsv1.Repository) {
+	w := tabwriter.NewWriter(os.Stdout, 0, 0, 2, ' ', 0)
+	fmt.Fprintf(w, "URL_PATTERN\tUSERNAME\tSSH_CREDS\tTLS_CREDS\n")
```
Hm, I was purposely using underscores so one could split lines using whitespace delimeter, i.e. for automation.

jannfis commented on pkg/apis/application/v1alpha1/types.go:
```diff
@@ -994,33 +994,45 @@ type Repository struct {
+// HasCredentials returns true when the receiver has been configured any credentials
 func (m *Repository) HasCredentials() bool {
-	return m.Username != "" || m.Password != "" || m.SSHPrivateKey != "" || m.InsecureIgnoreHostKey
+	return m.Username != "" || m.Password != "" || m.SSHPrivateKey != "" || m.TLSClientCertData != ""
```
Hm, seems so. Shall I open an issue for documentation purposes?

jannfis commented on pkg/apis/application/v1alpha1/types_test.go:
```diff
@@ -447,6 +452,88 @@ func TestRepository_HasCredentials(t *testing.T) {
+		repo Repository
+		want bool
+	}{
+		{
```
Thanks, I basically C&P'ed the tests. Maybe a general test refactoring would be nice one day, if time permits. They don't seem to follow any standards, alsi regarding `assert.Nil(t, err)` vs. `assert.NoError(t, err)` etc - probably organically grown over time :)

jannfis commented on server/repository/repository.go:
```diff
@@ -264,25 +323,71 @@ func (s *Server) Create(ctx context.Context, q *repositorypkg.RepoCreateRequest)
+	r := q.Repo
+
+	repo, err := s.db.CreateRepositoryCredentials(ctx, r)
+	if status.Convert(err).Code() == codes.AlreadyExists {
```
To be honest, I'm unsure. I C&P'd the code from the repository methods.

jannfis commented on server/repository/repository.go:
```diff
@@ -264,25 +323,71 @@ func (s *Server) Create(ctx context.Context, q *repositorypkg.RepoCreateRequest)
+	repo, err := s.db.CreateRepositoryCredentials(ctx, r)
+	if status.Convert(err).Code() == codes.AlreadyExists {
+		// act idempotent if existing spec matches new spec
+		existing, getErr := s.db.GetRepositoryCredentials(ctx, r.Repo)
```
C&P

jannfis commented on server/repository/repository.go:
```diff
@@ -264,25 +323,71 @@ func (s *Server) Create(ctx context.Context, q *repositorypkg.RepoCreateRequest)
+}
+
+// Delete removes a repository from the configuration
+// Deprecated: Use DeleteRepository() instead
```
Never accept defeat. We will fight the swagger daemon as we fight the lint monster. I think deprecating the old verb-only methods while parallel introducing new verb+noun methods is the way to go. It's a slow & cumbersome battle, but I'm confident we'll leave it as the winners.

jannfis commented on server/repository/repository.proto:
```diff
@@ -41,7 +41,10 @@ message RepoAppsResponse {
+	// Repo URL for query
 	string repo = 1;
+	// Whether to force a cache refresh on repo's connection state
+	bool forceRefresh = 2;
```
Hm, what would be the difference in the outcome? By default, results are read from the cache (if cached already), so `--refresh normal` would be redundant (i.e. is the default behaviour). If for UI consistency, I'd consider to change it, tho.

jannfis commented on docs/user-guide/private-repositories.md:
```diff
@@ -107,6 +107,53 @@ The Argo CD UI don't support configuring SSH credentials. The SSH credentials ca
+    Repositories can be added using CLI or Web UI without specifying credentials only after a matching credential has been set up
+
+!!! note
+    Matching credential template URL prefixes is done on a _best match_ effort, so the longest (best) match will take precedence. The order of definition is not important, as opposed to pre v1.4 configuration.
```
The problem with first match is, how to specify the order from UI (CLI or web) then? Ordering the entries in the YAML would be possible, but once you add entries via CLI or web UI, the control over the order is gone - except if you introduce weird switches like `--before <url>`, `--after <url>` or  `--position <num>`. Personally, I'd expect best match for patterns such as these.

alexec replied:
good point

alexec replied:
lets go with your algo

jannfis commented on server/repository/repository.proto:
```diff
@@ -64,21 +67,37 @@ message RepoAccessQuery {
+	// Whether to create in upsert mode
+	bool upsert = 2;
+	// Whether to operate on credential set instead of repository
+	bool credsOnly = 3;
```
Please see below

jannfis commented on util/db/repository.go:
```diff
@@ -200,6 +179,151 @@ func (db *db) DeleteRepository(ctx context.Context, repoURL string) error {
+		return nil, status.Errorf(codes.AlreadyExists, "repository credentials for '%s' already exists", r.Repo)
+	}
+
+	data := make(map[string][]byte)
```
C&P

jannfis replied:
Also, removed now

jannfis commented on cmd/argocd/commands/repo.go:
```diff
@@ -47,16 +47,19 @@ func NewRepoAddCommand(clientOpts *argocdclient.ClientOptions) *cobra.Command {
 		tlsClientCertPath              string
 		tlsClientCertKeyPath           string
 		enableLfs                      bool
+		credsOnly                      bool
```
Hm OK, probably makes sense. I can imagine this being a sub-command below `argocd repo`, such as `argocd repo creds`, for being more intuitive than `argocd creds`. Maybe I should also think about moving the `argocd cert` to `argocd repo cert`, but at some later point than v1.4 tho. Sounds more like something to introduce with v2.0.

alexec replied:
I'm kind of easy about this, I'd actually suggest `repocred` as the command, this is similar to `kubectl`

jannfis replied:
This came with the latest commit, so we now have a `repocreds` command and the functionality from `repo` cmd was removed.

jannfis commented:
Regarding the removal of `insecureIgnoreHostKey` from the credential templates, this was a design decision. It has a) not been well documented (no mention of this setting in the docs regarding repository credentials), and b) I see it as a potentially dangerous option, and also one that shouldn't be necessary anymore with the introduction of certificate management. The problem I see with it that it's a "fire & forget" option to open a potential huge security gap in the setup. It should be set using consent for each repository, if at all (i.e. must "hurt").

jannfis commented:
Great work on the Helm stuff, @alexec - I'll prepare a merge of master into this PR soon, and I suspect there is more conflicting stuff now that needs to be resolved. So it will probably take a few days to bring this PR into a mergeable state, considering that I have only little spare time to work on code the upcoming week.

@moensch I see that your issue has been solved already, so I won't consider any cherry-picked change from this PR, correct?

jannfis commented:
I have successfully merged the changes so far. The code builds, unit tests and e2e tests are all passing locally (except my very best e2e friend, `TestDeletingAppStuckInSync`). CircleCI seems to have some problems, fails with
```
#!/bin/bash -eo pipefail
rm -Rf /home/circleci/.go_workspace/src/github.com/argoproj/argo-cd

Received 'interrupt' signal
```
I have also seen this in other recent PR checks.

I have yet to conduct some manual tests and also to address some of the comments left in the review, so please don't consider this PR ready for merging yet.

jannfis commented:
@alexec I wonder with all the refactoring around repositories currently going on, whether I should wait until situation becomes stable or if we want to merge my changes early.

alexec commented:
> @alexec I wonder with all the refactoring around repositories currently going on, whether I should wait until situation becomes stable or if we want to merge my changes early.

I don't think you should wait. @alexmt ?

alexmt commented:
I think we should be ok.  @jannfis please go ahead with the changes.

alexec commented:
Don't forget to let us know when it is ready for review.

jannfis commented:
I'm still working on moving the credentials part out of `RepositoryService` to give the credentials a dedicated place to exist. This is more work than I expected, as I'm also giving it complete new data structures and types, so there'll be less confusion about what's actually a repository and what's a credential template.

jannfis commented:
I've introduced `RepoCredsService` with the latest commit, which handles CRUD operations for repository credentials. I have also changed semantics of `RepositoryCredentials` a little, so we now have a dedicated data structure for handling credential templates, as opposed to handle them using stub repository structures. This always introduced confusion of "what is a repository" and "what is a credential template", and which information should be where.

Final steps are now to adapt the UI to the new semantics, and also to merge the conflicts from master (I hope for the last time).

jannfis commented:
@alexec I think this is now finally ready for review. Please note that the change is now more invasive than initially was planned, but should be non breaking in terms of configuration - except when someone had `InsecureIgnoreHostKey` set in `repository.credentials` configuration (which I doubt, because it was not documented).

The UI still needs a little love, it's a little hackish implementation. I was thinking about moving the credentials part to its own settings page in the future, but for now, the current implementation in the repositories section seems good enough.

jannfis commented on cmd/argocd/commands/repo.go:
```diff
@@ -205,23 +261,37 @@ func printRepoUrls(repos []appsv1.Repository) {
 	}
 	command.Flags().StringVarP(&output, "output", "o", "wide", "Output format. One of: wide|url")
+	command.Flags().BoolVar(&credsOnly, "creds", false, "List repository credentials templates instead of connected repositories")
+	command.Flags().BoolVar(&forceRefresh, "force-refresh", false, "Force a cache refresh on connection status")
```
Followed the syntax - now there is `--refresh hard` for `repo list` as well.

alexec commented on cmd/argocd/commands/repocreds.go:
```diff
@@ -0,0 +1,197 @@
+			if sshPrivateKeyPath != "" {
+				if ok, _ := git.IsSSHURL(repo.URL); ok {
+					keyData, err := ioutil.ReadFile(sshPrivateKeyPath)
+					if err != nil {
```
minor - could use `error.Check(..)`

alexec commented on cmd/argocd/commands/repocreds.go:
```diff
@@ -0,0 +1,197 @@
+// Print the repository credentials as table
+func printRepoCredsTable(repos []appsv1.RepoCreds) {
+	w := tabwriter.NewWriter(os.Stdout, 0, 0, 2, ' ', 0)
+	fmt.Fprintf(w, "URL_PATTERN\tUSERNAME\tSSH_CREDS\tTLS_CREDS\n")
```
minor - if you want to provide parseable output, you could add a `-o yaml` option

alexec commented on pkg/apis/application/v1alpha1/types.go:
```diff
@@ -978,6 +978,22 @@ type ResourceActionParam struct {
+// RepoCreds holds a repository credentials definition
+type RepoCreds struct {
+	// URL is the URL that this credentials matches to
+	URL string `json:"url" protobuf:"bytes,1,opt,name=url"`
```
minor - do we want document how these are matched?

alexec commented on pkg/apis/application/v1alpha1/types_test.go:
```diff
@@ -405,7 +405,136 @@ func TestAppProjectSpec_DestinationClusters(t *testing.T) {
+			want: true,
+		},
+	}
+	for _, tt := range tests {
```
I think you could simply this test to two lines, as any field other than `EnableLFS`should not matter

alexec commented on server/repocreds/repocreds.go:
```diff
@@ -0,0 +1,114 @@
+	r := q.Creds
+
+	if r.URL == "" {
+		return nil, status.Errorf(codes.InvalidArgument, "must specify URL")
```
nice

alexec commented on server/repository/repository.go:
```diff
@@ -48,9 +48,14 @@ func NewServer(
+// repo and evaluate the results. Unless forceRefresh is set to true, the
+// result may be retrieved out of the cache.
+func (s *Server) getConnectionState(ctx context.Context, url string, forceRefresh bool) appsv1.ConnectionState {
+	if !forceRefresh {
```
Why do we need force refresh? I'm worried that we're using the wrong key for the correction state.

jannfis replied:
I think it'd be useful when you change repository credentials, you can see whether the change was successful or not instantly. Credentials aren't (can't be) checked for validity. 

The other possibility would be to invalidate the cache on connection state for every repository that matches the credentials' URL prefix, when the credentials change.

alexec commented on server/repository/repository.go:
```diff
@@ -267,11 +300,21 @@ func (s *Server) ValidateAccess(ctx context.Context, q *repositorypkg.RepoAccess
+		if err != nil {
+			return nil, err
+		}
+		if repoCreds != nil {
```
I think this func accepts nil? So no need to guard.

alexec commented:
This has been approved but needs to be sync'd with master. @jannfis do you want to do this?
